File: P4/ruby/SpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStationToUI'
require_relative 'ShotResult'
require_relative 'CardDealer'
require_relative 'Transformation'
require_relative 'SuppliesPackage'
require_relative 'Hangar'

module Deepspace
    class SpaceStation
        @@MAXFUEL = 100
        @@SHIELDLOSSPERUNITSHOT = 0.1

        def initialize (n, supplies, nMedals = 0, pendingDamage = nil, weapons = Array.new, shieldBoosters = Array.new, hangar = nil) # SpaceStation(n : string, supplies : SuppliesPackage)
			@name = n

			@ammoPower = supplies.ammoPower
            @fuelUnits = supplies.fuelUnits
			@shieldPower = supplies.shieldPower

			@nMedals = nMedals
			@pendingDamage = pendingDamage
			@weapons = weapons
			@shieldBoosters = shieldBoosters
			@hangar = hangar
		end
		
		attr_reader :name, :ammoPower, :fuelUnits, :shieldPower, :nMedals, :pendingDamage, :weapons, :shieldBoosters, :hangar


		def self.newCopy(station) # station : SpaceStation
			sup = Deepspace::SuppliesPackage.new(station.ammoPower, station.fuelUnits, station.shieldPower)
			new(station.name, sup, station.nMedals, station.pendingDamage.class.newCopy(station.pendingDamage), Array.new(station.weapons), Array.new(station.shieldBoosters), Hangar.newCopy(station.hangar))
		end

		def getUIversion
			SpaceStationToUI.new(self)
		end

		def to_s
			getUIversion.to_s 
		end

        def assignFuelValue (f)
        	@fuelUnits = f <= @@MAXFUEL ? f : @@MAXFUEL
        end

		def cleanPendingDamage
			if @pendingDamage != nil && @pendingDamage.hasNoEffect
				@pendingDamage = nil
			end
		end
		
		def receiveWeapon(w) # receiveWeapon(w : Weapon) : boolean
			if @hangar != nil
				@hangar.addWeapon(w) # true si tiene éxito, false si no
			else
				false
			end
		end
		
		def receiveShieldBooster(s)
			if @hangar != nil
				@hangar.addShieldBooster(s) # true si tiene éxito, false si no
			else
				false
			end
		end
		
		def receiveHangar(h)
			if @hangar.nil? 
				@hangar = h
			end
		end
		
		def discardHangar
			@hangar = nil
		end
		
		def receiveSupplies(s) # s: SuppliesPackage
			assignFuelValue(@fuelUnits + s.fuelUnits)
			@shieldPower += s.shieldPower
			@ammoPower += s.ammoPower
		end
		
		def setPendingDamage(d) # setPendingDamage(d : Damage) : void
			@pendingDamage = d.adjust(@weapons, @shieldBoosters)
		end

		def mountWeapon(i) # mountWeapon(i : int) : void
			if 0 <= i && i < @hangar.weapons.length
				new_weapon = @hangar.weapons.delete_at(i) unless hangar.nil?
				@weapons << new_weapon unless new_weapon.nil?
			end
		end

		def mountShieldBooster(i) # mountShieldBooster(i : int) : void
			if 0 <= i && i < @hangar.shieldBoosters.length
				new_shield_booster = @hangar.shieldBoosters.delete_at(i) unless hangar.nil?
				@shieldBoosters << new_shield_booster unless new_shield_booster.nil?
			end
		end

		def discardWeaponInHangar (i)
			if @hangar != nil && 0 <= i && i < @hangar.weapons.length
				@hangar.removeWeapon(i)
			end
		end

		def discardShieldBoosterInHangar (i) 
			if @hangar != nil && 0 <= i && i < @hangar.shieldBoosters.length
				@hangar.removeShieldBooster(i)
			end
		end

		def getSpeed
			@fuelUnits / @@MAXFUEL
		end

		def move
			@fuelUnits = @fuelUnits >= getSpeed ? @fuelUnits - getSpeed : 0
		end

		def validState
			@pendingDamage.nil? || @pendingDamage.hasNoEffect
		end

		def cleanUpMountedItems
			@weapons.delete_if {|x| x.uses == 0} unless @weapons.nil?
			@shieldBoosters.delete_if {|x| x.uses == 0} unless @shieldBoosters.nil?
		end

		def fire
			factor = 1.0
			@weapons.each {|w| factor *= w.useIt} unless @weapons.nil?
			@ammoPower*factor
		end

		def protection
			factor = 1.0
			@shieldBoosters.each {|s| factor *= s.useIt} unless @shieldBoosters.nil?
			@shieldPower*factor
		end

		def receiveShot(shot) # receiveShot(shot : float) : ShotResult
			myProtection = protection
			
			if myProtection >= shot
				@shieldPower = [0.0, @shieldPower - @@SHIELDLOSSPERUNITSHOT*shot].max
				ShotResult::RESIST
			else
				shieldPower = 0.0
				ShotResult::DONOTRESIST
			end
		end

		def setLoot(loot) # setLoot(loot : Loot) : Transformation
			dealer = CardDealer.instance
			h = loot.nHangars
			
			if h > 0
				hangar = dealer.nextHangar
				receiveHangar(hangar)
			end

			elements = loot.nSupplies
			for i in 1..elements
				sup = dealer.nextSuppliesPackage
				receiveSupplies(sup)
			end

			elements = loot.nWeapons
			for i in 1..elements
				weap = dealer.nextWeapon
				receiveWeapon(weap)
			end

			elements = loot.nShields
			for i in 1..elements
				sh = dealer.nextShieldBooster
				receiveShieldBooster(sh)
			end

			medals = loot.nMedals
			@nMedals += medals

			if loot.spaceCity
				Transformation::SPACECITY
			elsif loot.efficient
				Transformation::GETEFFICIENT
			else
				Transformation::NOTRANSFORM
			end
		end

		def discardWeapon(i) # discardWeapon(i : int) : void
			size = @weapons.length

			if i >= 0 && i < size
				w = @weapons.delete_at(i)
				if @pendingDamage != nil
					@pendingDamage.discardWeapon(w)
					cleanPendingDamage
				end
			end
		end

		def discardShieldBooster(i) # discardShieldBooster(i : int) : void
			size = @shieldBoosters.length

			if i >= 0 && i < size
				@shieldBoosters.delete_at(i)
				if @pendingDamage != nil
					@pendingDamage.discardShieldBooster	# He quitado el parámetro de discardShieldBooster
					cleanPendingDamage
				end
			end
		end

		# Especificación de acceso a métodos privados
		private :assignFuelValue, :cleanPendingDamage
    end
end

if $0 == __FILE__
=begin	
	require_relative 'SuppliesPackage'
	require_relative 'Weapon'
	require_relative 'WeaponType'
	require_relative 'Hangar'
	require_relative 'Damage'

	my_space_station = Deepspace::SpaceStation.new("Nombre de mi estación", Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station.to_s

	my_space_station.assignFuelValue(100)
	puts my_space_station.to_s

	puts
	d = Deepspace::Damage.newNumericWeapons(3,3)
	puts d.to_s
	my_space_station.setPendingDamage(d)
	puts my_space_station.to_s
	my_space_station.cleanPendingDamage
	puts my_space_station.to_s

	puts
	my_space_station.receiveHangar(Deepspace::Hangar.new(12))
	puts my_space_station.to_s
	my_space_station.receiveWeapon(Deepspace::Weapon.new(4, 5, 6))
	puts my_space_station.to_s
	my_space_station.receiveSupplies(Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station.to_s

	puts
	my_space_station.setPendingDamage(Deepspace::Damage.newSpecificWeapons(Array.new(3){Deepspace::WeaponType::MISSILE}, 1))
	puts my_space_station.to_s

	puts "\nCosa que quiero comprobar\n\n"
	puts my_space_station.mountWeapon(0)
	puts my_space_station.to_s

	puts my_space_station.mountShieldBooster(3)
	puts my_space_station.to_s

	puts my_space_station.discardWeaponInHangar(0)
	puts my_space_station.to_s
	puts my_space_station.discardShieldBoosterInHangar(5)
	puts my_space_station.to_s

	puts my_space_station.getSpeed
	puts my_space_station.move
	puts my_space_station.validState
	my_space_station.cleanUpMountedItems
	puts my_space_station.to_s

=end

	require_relative 'SuppliesPackage'
	require_relative 'Weapon'
	require_relative 'WeaponType'
	require_relative 'Hangar'
	require_relative 'Damage'

	my_space_station = Deepspace::SpaceStation.new("Nombre de mi estación", Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station

	puts
	my_space_station.receiveHangar(Deepspace::Hangar.new(12))
	puts my_space_station.to_s
	my_space_station.receiveWeapon(Deepspace::Weapon.new(4, Deepspace::WeaponType::LASER, 6))
	puts my_space_station.to_s
	my_space_station.receiveSupplies(Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station.to_s

	my_space_station.mountWeapon(0)
	puts my_space_station
	puts my_space_station.fire
	puts my_space_station.protection
	puts my_space_station.receiveShot(1.5)
	my_space_station.setLoot(Deepspace::Loot.new(1,1,1,1,1))
	puts my_space_station

	my_space_station.discardWeapon(0)
	my_space_station.discardShieldBooster(0)
end

File: Java/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.lang.Math;

public class SpaceStation implements SpaceFighter {
    private static int MAXFUEL = 100;
    private static float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private String name;
    
    private float ammoPower;
    private float fuelUnits;
    private float shieldPower;
    
    private int nMedals;
   
    private Damage pendingDamage;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    
    SpaceStation(String n, SuppliesPackage supplies){
        name = n;
        
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        
        nMedals = 0;
        
        pendingDamage = null;
        weapons = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
        hangar = null;
    }
    
    SpaceStation(String n, SuppliesPackage supplies, int nM, Damage pD, ArrayList<Weapon> w, ArrayList<ShieldBooster> sh, Hangar h) {
        name = n;
        
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        
        nMedals = nM;
        
        pendingDamage = pD;
        weapons = w;
        shieldBoosters = sh;
        hangar = h;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
    private void assignFuelValue(float f){
        fuelUnits = f <= MAXFUEL ? f : MAXFUEL;
    }
    
    private void cleanPendingDamage(){
        if(pendingDamage.hasNoEffect())
            pendingDamage = null;
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public String getName(){
        return name;
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public boolean receiveWeapon(Weapon w){
        if(hangar != null){
            return hangar.addWeapon(w);
        }else
            return false;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if(hangar != null){
            return hangar.addShielBooster(s);
        }else
            return false;
    }
    
    public void receiveHangar(Hangar h){
        if(hangar == null){
            hangar = h;
        }
    }
    
    public void discardHangar(){
        hangar = null;
    }
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        assignFuelValue(fuelUnits + s.getFuelUnits());
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public void mountWeapon(int i){
        if (0 <= i && i < hangar.getWeapons().size()) {
            if(hangar != null){
                Weapon newWeapon = hangar.getWeapons().remove(i);
                if(newWeapon != null)
                    weapons.add(newWeapon); 
            }
        }
    }
    
    public void mountShieldBooster(int i){
        if (0 <= i && i < hangar.getShieldBoosters().size()) {
            if(hangar != null){
                ShieldBooster newShield = hangar.getShieldBoosters().remove(i);
                if(newShield != null)
                    shieldBoosters.add(newShield);
            }
        }
    }
    
    public void discardWeaponInHangar(int i){
        if(hangar != null)
            hangar.removeWeapon(i);
    }
    
    public void discardShieldBoosterInHangar(int i){
        if(hangar != null)
            hangar.removeShieldBooster(i);
    }
    
    public float getSpeed(){
        return (float)(fuelUnits / MAXFUEL);
    }
    
    public void move(){
        fuelUnits = fuelUnits >= getSpeed() ? fuelUnits - getSpeed() : 0;
    }
    
    public boolean validState(){
        return (pendingDamage == null || pendingDamage.hasNoEffect());
    }
    
    public void cleanUpMountedItems(){
        if(weapons != null)
            for(int i=0; i<weapons.size(); ++i)
                if(weapons.get(i).getUses() == 0)
                    weapons.remove(i);
    
        if(shieldBoosters != null)
            for(int i=0; i<shieldBoosters.size(); ++i)
                if(shieldBoosters.get(i).getUses() == 0)
                    shieldBoosters.remove(i);
    }         
    
    public void discardWeapon(int i){
        int size = weapons.size();
        if(i >= 0 && i < size){
            Weapon w = weapons.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    
    public void discardShieldBooster(int i){
        int size = shieldBoosters.size();
        if(i >= 0 && i < size){
            ShieldBooster s = shieldBoosters.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }
    
    @Override
    public float fire() {
        float factor = 1;
        if (weapons != null) {
            for (Weapon w: weapons) {
                factor *= w.useIt();
            }
        }
        return ammoPower*factor;
    }
    
    @Override
    public float protection() {
        float factor = 1;
        if (shieldBoosters != null) {
            for (ShieldBooster s: shieldBoosters) {
                factor *= s.useIt();
            }
        }
        return shieldPower*factor;
    }
    
    @Override
    public ShotResult receiveShot(float shot) {
        float myProtection = protection();
        
        if (myProtection >= shot) {
            shieldPower = Math.max(0, shieldPower - SHIELDLOSSPERUNITSHOT*shot);
            return ShotResult.RESIST;
        }
        else {
            shieldPower = 0;
            return ShotResult.DONOTRESIST;
        }
    }
    
    public Transformation setLoot(Loot loot) {
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        
        if (h > 0) {
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        SuppliesPackage sup;
        for (int i=0; i<elements; i++) {
            sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        Weapon weap;
        for (int i=0; i<elements; i++) {
            weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        ShieldBooster sh;
        for (int i=0; i<elements; i++) {
            sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
        int medals = loot.getNMedals();
        nMedals += medals;
        
        Transformation trans;
        if (loot.spaceCity())
            trans = Transformation.SPACECITY;
        else if (loot.getEfficient())
            trans = Transformation.GETEFFICIENT;
        else
            trans = Transformation.NOTRANSFORM;
        
        return trans;
    }
}
